nodego/supervisor.go

```diff
@@ -0,0 +1,255 @@
+package nodego
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"runtime/debug"
+	"strconv"
+	"sync"
+	"time"
+
+	"golang.org/x/net/context/ctxhttp"
+)
+
+type logEntry struct {
+	TextPayload string
+	Severity    string
+	Time        string
+	ExecutionID string
+}
+
+func (e *logEntry) consoleOutput(executionID string) []byte {
+	var logBuf bytes.Buffer
+	fmt.Fprintf(&logBuf, "[%s]", e.Severity[:1])
+	fmt.Fprintf(&logBuf, "[%s]", time.Now().Format(isoTimeFormat))
+	if executionID != "" {
+		fmt.Fprintf(&logBuf, "[%s]", executionID)
+	}
+	logBuf.WriteByte(' ')
+	logBuf.WriteString(e.TextPayload)
+	if len(e.TextPayload) == 0 || e.TextPayload[len(e.TextPayload)-1] != '\n' {
+		logBuf.WriteByte('\n')
+	}
+	return logBuf.Bytes()
+}
+
+type logBatch struct {
+	Entries []*logEntry
+
+	payloadLength int
+	ready         chan struct{}
+}
+
+// addEntry adds a log entry to the batch.
+//
+// Note: addEntry is not thread safe.
+func (batch *logBatch) addEntry(entry *logEntry) {
+	if batch.Entries == nil {
+		close(batch.ready)
+	}
+
+	batch.Entries = append(batch.Entries, entry)
+	batch.payloadLength += len(entry.TextPayload)
+}
+
+var loggingContext struct {
```

Comment:
Optional: What you have now is fine, but what I was envisioning was a named type so the following can happen:
logBatchReportWorker becomes a method on the new named type.
startNewLogBatch becomes a method on the new named type.
loggerMiddleware calls a setFunctionExecutionID method on the new named type instead of reaching into its internals.
The locked logic in supervisorWriter.Write moves to a method on the new named type. All of that logic is more tied into the log management than actually creating the log data.